### DISCUSSION TITLE
How to explain a contrib/custom field type to PHPstan

### DISCUSSION THREAD

**Discussion by mxr576** [mglaman/phpstan-drupal#473] [Q&A] [answered]
I have a custom field type defined on a project, but I guess the issue exists for contrib field types as well.

Since the source code is under my control, I thought adding a `@property string $foo` is enough to explain PHPStan that `foo` exists in the `$entity->get('field_bar')->foo` and the return type is a string, but that did not work. I also tried to use the stubbing approach that PHPStanDrupal does to fix Drupal sources, but that also did not work... I started to debug and realized that every `$entity->get('field_bar')` has dumped type `Dumped type: Drupal\Core\Field\FieldItemListInterface<Drupal\Core\Field\FieldItemInterface> `.  I am not sure how PHPStan can be taught to understand `$entity->get('field_bar')` is a `Drupal\my_module\Plugin\Field\FieldType\FooType` but if I am right than PHPStan Drupal also does not even try that at this moment.

I have checked what [EntityFieldsViaMagicReflectionExtension.php](https://github.com/mglaman/phpstan-drupal/blob/480245d5d0bd1858e01c43d738718dab85be0ad2/src/Reflection/EntityFieldsViaMagicReflectionExtension.php) and [EntityFieldReflection.php]( https://github.com/mglaman/phpstan-drupal/blob/480245d5d0bd1858e01c43d738718dab85be0ad2/src/Reflection/EntityFieldReflection.php) are doing and basically copied them to register the return type of my custom property...

But is the right and best way to do this? Seems ugly and utterly over complicated :) What do I miss?

**Comment by mglaman** [comment, marked as answer]
Right now there is no way unless you're extremely verbose

```
$field_item = $entity->get('field_bar')->first();
assert($field_item instanceof MyFieldItem);
```

A goal of https://github.com/mglaman/phpstan-drupal/issues/343 was to allow mapping fields to entity types. I guess this could happen as configuration without that! So the feature would be a field mapping alongside the entity mapping to say "this field has X field item list class and Y field item class"
